### ISSUE TITLE
autogeneration with conditional indices

### ISSUE THREAD

**Issue by chriswhite199** [sqlalchemy/alembic#1491]
**Describe the bug**

Using conditional Indices with `ddl_if`, alembic ignores the target dialect and creates all the indices, not just the index that matches the conditional constraint.

```python3
from sqlalchemy import Index, String
from sqlalchemy.orm import DeclarativeBase, Mapped, mapped_column


class Base(DeclarativeBase):
    pass


class User(Base):
    __tablename__ = "user"

    __table_args__ = (
        Index("my_pg_index_psql", "name").ddl_if(dialect="postgresql"),
        Index("my_pg_index_sqlite", "name").ddl_if(dialect="sqlite"),
    )

    id: Mapped[int] = mapped_column(primary_key=True)
    name: Mapped[str] = mapped_column(String(30))
```

```bash
./bin/alembic revision --autogenerate -m 'DB init'
INFO  [alembic.runtime.migration] Context impl PostgresqlImpl.
INFO  [alembic.runtime.migration] Will assume transactional DDL.
INFO  [alembic.autogenerate.compare] Detected added table 'user_account'
INFO  [alembic.autogenerate.compare] Detected added index ''my_pg_index_psql'' on '('name',)'
INFO  [alembic.autogenerate.compare] Detected added index ''my_pg_index_sqlite'' on '('name',)'
```

Generate python file contains two create_index statements, should be a single stmt:

```python3
def upgrade() -> None:
    # ### commands auto generated by Alembic - please adjust! ###
    op.create_table('user',
    sa.Column('id', sa.Integer(), nullable=False),
    sa.Column('name', sa.String(length=30), nullable=False),
    sa.PrimaryKeyConstraint('id')
    )
    op.create_index('my_pg_index_psql', 'user_account', ['name'], unique=False)
    op.create_index('my_pg_index_sqlite', 'user_account', ['name'], unique=False)
    # ### end Alembic commands ###
```

**Expected behavior**
Only the index for the target dialect should be created, not both. Note if you use the same index name for both Index defs, you will get a single op.create_index but the configuration will be random.

**To Reproduce**
As above, using postgres as the db in alembic.ini:

```python
sqlalchemy.url = postgresql://postgres:password@localhost/postgres
```

**Versions.**
 - OS: Linux
 - Python: 3.12.2
 - Alembic: 1.13.1
 - SQLAlchemy: 2.0.31
 - Database: Postgres 14
 - DBAPI: ?

**Comment by zzzeek** [comment]
hi -

this is very special use case so you would need to manually adjust the autogenerated file for now.   logic would need to be added to autogenerate to look for these ddl_ifs but it would likely then not be able to correctly generate each dialect-specific index, since autogen is only against one database.

**Comment by chriswhite199** [comment]
Fair enough, just checking whether this was a known issue, bug, not supported etc.

For the case of autogen, it would logically follow that it should be easier because it's a single dialect, so just ignore those not for the current dialect?

I'll see if i can understand the autogen code and maybe submit a PR for this.

Might consider updating https://alembic.sqlalchemy.org/en/latest/autogenerate.html#what-does-autogenerate-detect-and-what-does-it-not-detect to denote this edge case

**Comment by zzzeek** [comment]
well it would render the Index() for just one of the databases, and not the other, meaning your migration file will only work on that one database.  that is, it wouldnt generate a cross-compatible migration file even though your model is cross compatible.

**Comment by chriswhite199** [comment]
So for my specific use case, the indexing options are different for PSQL vs SQLite, and SQLite is only used for quick integration tests. I only ever autogenerate for PSQL.

Would you also not be able to follow this guidance and just generate multiple migration file sets for each dialect type:

https://alembic.sqlalchemy.org/en/latest/cookbook.html#run-multiple-alembic-environments-from-one-ini-file

**Comment by zzzeek** [comment]
if you really wanted multiple separate environments, sure, usually people want one migration set for a certain kind of schema and they use cross-compatibility features for things like special indexes.

anyway, PRs to improve the behavior are welcome as long as you can make some tests for them

**Comment by CaselIT** [comment]
how could a direct support api look like? pass a kw arg with a dict of kw similar to https://docs.sqlalchemy.org/en/20/core/constraints.html#sqlalchemy.schema.HasConditionalDDL.ddl_if ?
```py
op.create_index('my_pg_index_psql', 'user_account', ['name'], unique=False, ddl_if={'dialect': 'postgresql')
op.create_index('my_pg_index_sqlite', 'user_account', ['name'], unique=False ddl_if={'dialect': 'sqlite')
```

something else like a conditional?
```py
if op.ddl_if(dialect='postgresql'):
  op.create_index('my_pg_index_psql', 'user_account', ['name'], unique=False)
if op.ddl_if(dialect='sqlite'):
  op.create_index('my_pg_index_sqlite', 'user_account', ['name'], unique=False)
```

**Comment by zzzeek** [comment]
I like the second form better, but it might be more challenging from a python rendering perspective (or it likely doesnt matter much).     I'm pretty sure we talked about, or I had in my head at some point, that we should have some formal "if <dialect in use>" construct in alembic (annnd....I'm pretty sure we dont yet, but not 100% sure)

**Comment by CaselIT** [comment]
it's for sure more general, since it could be used by user to make conditional sections.
It's likely easy to add as an api, but supporting it in the autogenerate it may be a bit more complex

**Comment by MattOates** [comment]
This feels like an important feature for post SQLAlchemy 2.0. I just ran into this problem and solved with some logic in env.py looking at dialects for rendering migrations. Im in the simpler setup of each DB dialect renders a distinct migration lineage so just needed to prevent the wrong dialect pieces being selected at all. Im running into a fair few bits like this where it just feels like alembic is lagging behind SQLAlchemy's status. If I spent some time specifically looking at this and @CaselIT second example, would a PR be welcome?

**Comment by zzzeek** [comment]
we welcome PRs for things like this if you run by us what you want to do first :)   , so that we aren't rejecting a big tract of work you did.   i havent looked at this issue in 18 months so I'd have to re-familiarize with it.      overall the class of issues that's "autogenerate needs to add things that are highly DB specific" is pretty much unsolved in alembic, the PG ENUM case being the biggest one.   we dont have any usual way to approach this problem for autogen.

**Comment by MattOates** [comment]
This is specifically the case where someone has marked an index with `Index.ddl_if(dialect="some dialect name")` and then with a normal create all in SQLModel/SQLAlchemy if you're not the specified dialect then this index or constraint is ignored. You can get the same behaviour simply enough in Alembic with something like the following in your env.py . So it could be left as that.

```python
 # Skip dialect-specific indexes that don't match the current dialect
    if type_ == "index":
        # Check for dialect-specific parameters (e.g., postgresql_using, mysql_using, etc.)
        if hasattr(obj, "kwargs"):
            for key in obj.kwargs.keys():
                if key.endswith("_using") or key.endswith("_where"):
                    # Extract the dialect name from keys like "postgresql_using"
                    index_dialect = key.split("_")[0]
                    # Normalize dialect names (spanner+spanner -> spanner)
                    current_dialect = dialect.split("+")[0] if "+" in dialect else dialect
                    if index_dialect != current_dialect:
                        return False
        
        # Check for ddl_if dialect restrictions
        if hasattr(obj, "dialect_options"):
            for dialect_name, options in obj.dialect_options.items():
                if "ddl_if" in options:
                    # Normalize dialect names
                    current_dialect = dialect.split("+")[0] if "+" in dialect else dialect
                    if dialect_name != current_dialect:
                        return False
``` 

So I guess the simplest PR might just be some docs in the cookbook https://alembic.sqlalchemy.org/en/latest/cookbook.html for starters... I've been doing a lot of things pushing where alembic is at in the past few months in a codebase where I have spanner, bigquery and postgresql all migrating from the same model defs. There is a huge amount of complexity in my env.py including an entire effort to attempt to do things as lame as convert uniqe constraints into unique indexes for Spanner only etc. The absolute minimum for this project is ensuring index definitions for special index types dont just render out into the migrations for the wrong database tech.

**Comment by zzzeek** [comment]
the cookbook is exactly where we develop these new use cases where some of them become real features.  so if you have a cookbook approach, that's an easy PR for us to accept
